source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,496 @@
+      // The entries in the scheduled_services are already put on the apple dispatch queue. It would be a bad memory
+      // access if we destroy the entries here. We instead setting a cancel flag to cancel the task when the
+      // dispatch_queue execute the entry.
+      struct aws_linked_list_node *iter = NULL;
+      for (iter = aws_linked_list_begin(&dispatch_loop->synced_data.scheduling_state.scheduled_services);
+           iter != aws_linked_list_end(&dispatch_loop->synced_data.scheduling_state.scheduled_services);
+           iter = aws_linked_list_next(iter)) {
+          struct scheduled_service_entry *entry = AWS_CONTAINER_OF(iter, struct scheduled_service_entry, node);
+          entry->cancel = true;
+      }
+      dispatch_loop->synced_data.suspended = true;
+      dispatch_loop->synced_data.is_executing = false;
+      aws_mutex_unlock(&dispatch_loop->synced_data.lock);
+    });
+
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Releasing Dispatch Queue.", (void *)event_loop);
+    aws_ref_count_release(&dispatch_loop->ref_count);
+}
+
+static int s_wait_for_stop_completion(struct aws_event_loop *event_loop) {
+    (void)event_loop;
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_run(struct aws_event_loop *event_loop) {
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    aws_mutex_lock(&dispatch_loop->synced_data.lock);
+    if (dispatch_loop->synced_data.suspended) {
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Starting event-loop thread.", (void *)event_loop);
+        dispatch_resume(dispatch_loop->dispatch_queue);
+        dispatch_loop->synced_data.suspended = false;
+    }
+    aws_mutex_unlock(&dispatch_loop->synced_data.lock);
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_stop(struct aws_event_loop *event_loop) {
+    struct dispatch_loop *dispatch_loop = event_loop->impl_data;
+
+    aws_mutex_lock(&dispatch_loop->synced_data.lock);
+    if (!dispatch_loop->synced_data.suspended) {
+        dispatch_loop->synced_data.suspended = true;
+        AWS_LOGF_INFO(AWS_LS_IO_EVENT_LOOP, "id=%p: Stopping event-loop thread.", (void *)event_loop);
+        // Suspend will increase the dispatch reference count. It is required to call resume before
+        // releasing the dispatch queue.
+        dispatch_suspend(dispatch_loop->dispatch_queue);
+    }
+    aws_mutex_unlock(&dispatch_loop->synced_data.lock);
+
+    return AWS_OP_SUCCESS;
+}
+
+void try_schedule_new_iteration(struct aws_event_loop *loop, uint64_t timestamp);
+
+// returns true if we should execute an iteration, false otherwise
+bool begin_iteration(struct scheduled_service_entry *entry) {
+    bool should_execute_iteration = false;
+    struct dispatch_loop *dispatch_loop = entry->loop->impl_data;
+
+    aws_mutex_lock(&dispatch_loop->synced_data.lock);
+
+    // someone else is already going, do nothing
+    if (dispatch_loop->synced_data.scheduling_state.is_executing_iteration) {
```

Comment:
I think I'm missing something. `is_executing_iteration` is set to true only within `run_iteration`. Does this check mean that multiple `run_iteration` can run in parallel?